lib/cn/node_split.c

```diff
@@ -8,38 +8,53 @@
 #include <stdint.h>
 #include <stdlib.h>
 
+#include <hse_ikvdb/kvset_view.h>
 #include <hse_util/element_source.h>
 #include <hse_util/event_counter.h>
 #include <hse_util/hse_err.h>
+#include <hse_util/assert.h>
 #include <hse_util/keycmp.h>
 
 #include "cn_tree_internal.h"
 #include "kvset.h"
 #include "kvset_internal.h"
 
 struct reverse_kblk_iterator {
-    struct kvset *oi_ks;
-    struct element_source oi_es;
-    bool oi_eof;
-    uint32_t oi_offset;
+    struct kvset *ks;
+    struct element_source es;
+    bool eof;
+    uint32_t offset;
+};
+
+struct forward_wbt_leaf_iterator {
+    struct kvset *ks;
+    struct element_source es;
+    bool eof;
+    struct {
+        uint32_t kblk_idx;
+        uint32_t leaf_idx;
+    } offset;
 };
 
 static bool
-next(struct element_source *source, void **data)
+reverse_kblk_iterator_next(struct element_source *source, void **data)
 {
-    struct reverse_kblk_iterator *iter = container_of(source, struct reverse_kblk_iterator, oi_es);
+    struct reverse_kblk_iterator *iter = container_of(source, struct reverse_kblk_iterator, es);
+
+    INVARIANT(source);
+    INVARIANT(data);
 
-    if (iter->oi_eof)
+    if (iter->eof)
         return false;
 
-    assert(iter->oi_offset > 0);
+    assert(iter->offset > 0);
 
-    *data = &iter->oi_ks->ks_kblks[--iter->oi_offset];
+    *data = &iter->ks->ks_kblks[--iter->offset];
 
-    if (iter->oi_offset == 0)
-        iter->oi_eof = true;
+    if (iter->offset == 0)
+        iter->eof = true;
 
-    return iter->oi_eof;
+    return iter->eof;
```

Comment:
This returns EOF on element too early.   Seems like the logic should be something like:
```
if (iter->offset == 0)
  return eof;

*data = &iter->ks->ks_kblks[--iter->offset];
return false;

```

Reply:
what i wrote isn't right either. :(

Reply:
Same issue for other iterator it seems. I will fix that.